File: app/src/main/java/ru/p3tr0vich/fuel/FragmentBackup.java
package ru.p3tr0vich.fuel;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import ru.p3tr0vich.fuel.helpers.DatabaseBackupXmlHelper;
import ru.p3tr0vich.fuel.utils.Utils;
import ru.p3tr0vich.fuel.utils.UtilsLog;

import static android.support.v4.content.ContextCompat.checkSelfPermission;

public class FragmentBackup extends FragmentBase {

    private static final String TAG = "FragmentBackup";

    private static final int REQUEST_CODE_DIALOG_PROGRESS = 100;
    private static final int REQUEST_CODE_DIALOG_QUESTION = 200;

    private static final int REQUEST_CODE_PERMISSION_WRITE_EXTERNAL_STORAGE = 300;
    private static final int REQUEST_CODE_PERMISSION_READ_EXTERNAL_STORAGE = 301;

    private final DatabaseBackupXmlHelper mDatabaseBackupXmlHelper = new DatabaseBackupXmlHelper();

    @Override
    public int getTitleId() {
        return R.string.title_backup;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        FragmentManager fragmentManager = getFragmentManager();

        assert fragmentManager != null;

        FragmentDialogProgress fragmentDialogProgress =
                (FragmentDialogProgress) fragmentManager.findFragmentByTag(FragmentDialogProgress.TAG);
        FragmentDialogQuestion fragmentDialogQuestion =
                (FragmentDialogQuestion) fragmentManager.findFragmentByTag(FragmentDialogQuestion.TAG);

        UtilsLog.d(TAG, "onCreate", "fragmentDialogProgress " +
                (fragmentDialogProgress == null ? "=" : "!") + "= null");

        if (fragmentDialogProgress != null)
            fragmentDialogProgress.setTargetFragment(this, REQUEST_CODE_DIALOG_PROGRESS);
        if (fragmentDialogQuestion != null)
            fragmentDialogQuestion.setTargetFragment(this, REQUEST_CODE_DIALOG_QUESTION);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        UtilsLog.d(TAG, "onCreateView");

        View view = inflater.inflate(R.layout.fragment_backup, container, false);

        ((TextView) view.findViewById(R.id.text_directory)).setText(
                mDatabaseBackupXmlHelper.getExternalDirectory().toString());
        ((TextView) view.findViewById(R.id.text_file)).setText(
                mDatabaseBackupXmlHelper.getFileName().toString());

        view.findViewById(R.id.btn_save).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                btnSaveClick();
            }
        });

        view.findViewById(R.id.btn_load).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                btnLoadClick();
            }
        });

        return view;
    }

    private void startSave() {
        UtilsLog.d(TAG, "startSave");

        FragmentDialogProgress.show(this, REQUEST_CODE_DIALOG_PROGRESS, mDatabaseBackupXmlHelper, true);
    }

    private void startLoad() { // TODO: Сохранять старые в old?
        UtilsLog.d(TAG, "startLoad");

        FragmentDialogProgress.show(this, REQUEST_CODE_DIALOG_PROGRESS, mDatabaseBackupXmlHelper, false);
    }

    private void stopOperationXml(@DatabaseBackupXmlHelper.BackupResult int result) {
        String resultMessage;

        switch (result) {
            case DatabaseBackupXmlHelper.RESULT_SAVE_OK:
                resultMessage = "File '" + mDatabaseBackupXmlHelper.getFileName() + "' save in '" +
                        mDatabaseBackupXmlHelper.getExternalDirectory() + "' without errors";
                break;
            case DatabaseBackupXmlHelper.RESULT_LOAD_OK:
                resultMessage = "File '" + mDatabaseBackupXmlHelper.getFileName() + "' load from '" +
                        mDatabaseBackupXmlHelper.getExternalDirectory() + "' without errors";
                break;
            case DatabaseBackupXmlHelper.RESULT_ERROR_MKDIRS:
                resultMessage = getString(R.string.message_error_mkdirs, mDatabaseBackupXmlHelper.getExternalDirectory());
                break;
            case DatabaseBackupXmlHelper.RESULT_ERROR_CREATE_XML:
                resultMessage = getString(R.string.message_error_create_xml);
                break;
            case DatabaseBackupXmlHelper.RESULT_ERROR_CREATE_FILE:
                resultMessage = getString(R.string.message_error_create_file,
                        mDatabaseBackupXmlHelper.getFileName(),
                        mDatabaseBackupXmlHelper.getExternalDirectory());
                break;
            case DatabaseBackupXmlHelper.RESULT_ERROR_SAVE_FILE:
                resultMessage = getString(R.string.message_error_save_file,
                        mDatabaseBackupXmlHelper.getFileName(),
                        mDatabaseBackupXmlHelper.getExternalDirectory());
                break;
            case DatabaseBackupXmlHelper.RESULT_ERROR_DIR_NOT_EXISTS:
                resultMessage = getString(R.string.message_error_dir_not_exists, mDatabaseBackupXmlHelper.getExternalDirectory());
                break;
            case DatabaseBackupXmlHelper.RESULT_ERROR_FILE_NOT_EXISTS:
                resultMessage = getString(R.string.message_error_file_not_exists,
                        mDatabaseBackupXmlHelper.getFileName(),
                        mDatabaseBackupXmlHelper.getExternalDirectory());
                break;
            case DatabaseBackupXmlHelper.RESULT_ERROR_READ_FILE:
                resultMessage = getString(R.string.message_error_read_file);
                break;
            case DatabaseBackupXmlHelper.RESULT_ERROR_PARSE_XML:
                resultMessage = getString(R.string.message_error_parse_xml);
                break;
            default:
                return;
        }

        UtilsLog.d(TAG, "stopOperationXml: " + resultMessage);

        if (result == DatabaseBackupXmlHelper.RESULT_SAVE_OK) {
            Utils.toast(R.string.message_save_file_ok);
        } else if (result == DatabaseBackupXmlHelper.RESULT_LOAD_OK) {
            Utils.toast(R.string.message_load_file_ok);

            preferencesHelper.putFullSync(true);
        } else {
            FragmentActivity activity = getActivity();

            assert activity != null;

            FragmentDialogMessage.show(activity, getString(R.string.title_message_error), resultMessage);
        }
    }

    private boolean checkPermission(final int permissionCode) {
        if (Build.VERSION.SDK_INT < 23) {
            return true;
        }

        Context context = getContext();

        assert context != null;

        final String permission;

        switch (permissionCode) {
            case REQUEST_CODE_PERMISSION_WRITE_EXTERNAL_STORAGE:
                permission = Manifest.permission.WRITE_EXTERNAL_STORAGE;
                break;
            case REQUEST_CODE_PERMISSION_READ_EXTERNAL_STORAGE:
                permission = Manifest.permission.READ_EXTERNAL_STORAGE;
                break;
            default:
                return false;
        }

        if (checkSelfPermission(context, permission) == PackageManager.PERMISSION_GRANTED) {
            return true;
        }

        requestPermissions(new String[]{permission}, permissionCode);

        return false;
    }

    private void btnSaveClick() {
        if (checkPermission(REQUEST_CODE_PERMISSION_WRITE_EXTERNAL_STORAGE)) {
            startSave();
        }
    }

    private void startLoadQuestion() {
        FragmentDialogQuestion.show(this, REQUEST_CODE_DIALOG_QUESTION,
                R.string.dialog_caption_load_from_xml,
                preferencesHelper.isSyncEnabled() ?
                        R.string.message_dialog_load_from_xml_sync :
                        R.string.message_dialog_load_from_xml, R.string.dialog_btn_load, R.string.dialog_btn_disagree);
    }

    private void btnLoadClick() {
        if (checkPermission(REQUEST_CODE_PERMISSION_READ_EXTERNAL_STORAGE)) {
            startLoadQuestion();
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        UtilsLog.d(TAG, "onActivityResult");

        if (resultCode != Activity.RESULT_OK) {
            return;
        }

        switch (requestCode) {
            case REQUEST_CODE_DIALOG_PROGRESS:
                stopOperationXml(FragmentDialogProgress.getResult(data));
                break;
            case REQUEST_CODE_DIALOG_QUESTION:
                startLoad();
                break;
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            switch (requestCode) {
                case REQUEST_CODE_PERMISSION_WRITE_EXTERNAL_STORAGE:
                    startSave();
                    break;
                case REQUEST_CODE_PERMISSION_READ_EXTERNAL_STORAGE:
                    startLoadQuestion();
            }
        } else {
            FragmentActivity activity = getActivity();

            assert activity != null;

            FragmentDialogMessage.show(activity, R.string.title_message_error, R.string.message_need_permission_to_storage);
        }
    }
}

File: app/src/main/kotlin/ru/p3tr0vich/fuel/fragments/FragmentBackup.kt
package ru.p3tr0vich.fuel.fragments

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.support.v4.content.ContextCompat.checkSelfPermission
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import ru.p3tr0vich.fuel.R
import ru.p3tr0vich.fuel.factories.FragmentFactory
import ru.p3tr0vich.fuel.helpers.DatabaseBackupXmlHelper
import ru.p3tr0vich.fuel.utils.Utils
import ru.p3tr0vich.fuel.utils.UtilsLog

class FragmentBackup : FragmentBase(FragmentFactory.Ids.BACKUP) {

    private val databaseBackupXmlHelper = DatabaseBackupXmlHelper()

    override val titleId: Int
        get() = R.string.title_backup

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val fragmentDialogProgress = fragmentManager?.findFragmentByTag(FragmentDialogProgress.TAG) as FragmentDialogProgress?
        val fragmentDialogQuestion = fragmentManager?.findFragmentByTag(FragmentDialogQuestion.TAG) as FragmentDialogQuestion?

        UtilsLog.d(TAG, "onCreate", "fragmentDialogProgress = $fragmentDialogProgress")

        fragmentDialogProgress?.setTargetFragment(this, REQUEST_CODE_DIALOG_PROGRESS)
        fragmentDialogQuestion?.setTargetFragment(this, REQUEST_CODE_DIALOG_QUESTION)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        UtilsLog.d(TAG, "onCreateView")

        with(inflater.inflate(R.layout.fragment_backup, container, false)) {
            (findViewById<View>(R.id.text_directory) as TextView).text =
                    databaseBackupXmlHelper.externalDirectory.toString()
            (findViewById<View>(R.id.text_file) as TextView).text =
                    databaseBackupXmlHelper.fileName.toString()

            findViewById<View>(R.id.btn_save).setOnClickListener { btnSaveClick() }

            findViewById<View>(R.id.btn_load).setOnClickListener { btnLoadClick() }

            return this
        }
    }

    private fun startSave() {
        UtilsLog.d(TAG, "startSave")

        FragmentDialogProgress.show(this, REQUEST_CODE_DIALOG_PROGRESS, databaseBackupXmlHelper, true)
    }

    private fun startLoad() { // TODO: Сохранять старые в old?
        UtilsLog.d(TAG, "startLoad")

        FragmentDialogProgress.show(this, REQUEST_CODE_DIALOG_PROGRESS, databaseBackupXmlHelper, false)
    }

    private fun stopOperationXml(@DatabaseBackupXmlHelper.BackupResult result: Int) {
        val resultMessage: String = when (result) {
            DatabaseBackupXmlHelper.RESULT_SAVE_OK -> "File '${databaseBackupXmlHelper.fileName}' save in '${databaseBackupXmlHelper.externalDirectory}' without errors"
            DatabaseBackupXmlHelper.RESULT_LOAD_OK -> "File '${databaseBackupXmlHelper.fileName}' load from '${databaseBackupXmlHelper.externalDirectory}' without errors"
            DatabaseBackupXmlHelper.RESULT_ERROR_MKDIRS -> getString(R.string.message_error_mkdirs, databaseBackupXmlHelper.externalDirectory)
            DatabaseBackupXmlHelper.RESULT_ERROR_CREATE_XML -> getString(R.string.message_error_create_xml)
            DatabaseBackupXmlHelper.RESULT_ERROR_CREATE_FILE -> getString(R.string.message_error_create_file,
                    databaseBackupXmlHelper.fileName,
                    databaseBackupXmlHelper.externalDirectory)
            DatabaseBackupXmlHelper.RESULT_ERROR_SAVE_FILE -> getString(R.string.message_error_save_file,
                    databaseBackupXmlHelper.fileName,
                    databaseBackupXmlHelper.externalDirectory)
            DatabaseBackupXmlHelper.RESULT_ERROR_DIR_NOT_EXISTS -> getString(R.string.message_error_dir_not_exists, databaseBackupXmlHelper.externalDirectory)
            DatabaseBackupXmlHelper.RESULT_ERROR_FILE_NOT_EXISTS -> getString(R.string.message_error_file_not_exists,
                    databaseBackupXmlHelper.fileName,
                    databaseBackupXmlHelper.externalDirectory)
            DatabaseBackupXmlHelper.RESULT_ERROR_READ_FILE -> getString(R.string.message_error_read_file)
            DatabaseBackupXmlHelper.RESULT_ERROR_PARSE_XML -> getString(R.string.message_error_parse_xml)
            else -> return
        }

        UtilsLog.d(TAG, "stopOperationXml: $resultMessage")

        when (result) {
            DatabaseBackupXmlHelper.RESULT_SAVE_OK -> {
                Utils.toast(R.string.message_save_file_ok)
            }
            DatabaseBackupXmlHelper.RESULT_LOAD_OK -> {
                Utils.toast(R.string.message_load_file_ok)

                preferencesHelper.isFullSync = true
            }
            else -> {
                FragmentDialogMessage.show(activity!!, getString(R.string.title_message_error), resultMessage)
            }
        }
    }

    private fun checkPermission(permissionCode: Int): Boolean {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            return true
        }

        val permission: String = when (permissionCode) {
            REQUEST_CODE_PERMISSION_WRITE_EXTERNAL_STORAGE -> Manifest.permission.WRITE_EXTERNAL_STORAGE
            REQUEST_CODE_PERMISSION_READ_EXTERNAL_STORAGE -> Manifest.permission.READ_EXTERNAL_STORAGE
            else -> return false
        }

        if (checkSelfPermission(context!!, permission) == PackageManager.PERMISSION_GRANTED) {
            return true
        }

        requestPermissions(arrayOf(permission), permissionCode)

        return false
    }

    private fun btnSaveClick() {
        if (checkPermission(REQUEST_CODE_PERMISSION_WRITE_EXTERNAL_STORAGE)) {
            startSave()
        }
    }

    private fun startLoadQuestion() {
        FragmentDialogQuestion.show(this, REQUEST_CODE_DIALOG_QUESTION,
                R.string.dialog_caption_load_from_xml,
                if (preferencesHelper.isSyncEnabled)
                    R.string.message_dialog_load_from_xml_sync
                else
                    R.string.message_dialog_load_from_xml, R.string.dialog_btn_load, R.string.dialog_btn_disagree)
    }

    private fun btnLoadClick() {
        if (checkPermission(REQUEST_CODE_PERMISSION_READ_EXTERNAL_STORAGE)) {
            startLoadQuestion()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        UtilsLog.d(TAG, "onActivityResult")

        if (resultCode != Activity.RESULT_OK) {
            return
        }

        when (requestCode) {
            REQUEST_CODE_DIALOG_PROGRESS -> stopOperationXml(FragmentDialogProgress.getResult(data!!))
            REQUEST_CODE_DIALOG_QUESTION -> startLoad()
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            when (requestCode) {
                REQUEST_CODE_PERMISSION_WRITE_EXTERNAL_STORAGE -> startSave()
                REQUEST_CODE_PERMISSION_READ_EXTERNAL_STORAGE -> startLoadQuestion()
            }
        } else {
            FragmentDialogMessage.show(activity!!, R.string.title_message_error, R.string.message_need_permission_to_storage)
        }
    }

    companion object {

        private const val TAG = "FragmentBackup"

        private const val REQUEST_CODE_DIALOG_PROGRESS = 100
        private const val REQUEST_CODE_DIALOG_QUESTION = 200

        private const val REQUEST_CODE_PERMISSION_WRITE_EXTERNAL_STORAGE = 300
        private const val REQUEST_CODE_PERMISSION_READ_EXTERNAL_STORAGE = 301
    }
}